js/packages/mobile-wallet-adapter-protocol/android/src/main/java/com/solanamobile/mobilewalletadapter/reactnative/SolanaMobileWalletAdapterModule.kt

```diff
@@ -0,0 +1,210 @@
+                    key,
+                    readableMap.getArray(key)?.let { convertArrayToJson(it) }
+                )
+            }
+        }
+        return json
+    }
+
+    @Throws(JSONException::class)
+    private fun convertArrayToJson(readableArray: ReadableArray): JSONArray {
+        val array = JSONArray()
+        for (i in 0 until readableArray.size()) {
+            when (readableArray.getType(i)) {
+                ReadableType.Null -> {}
+                ReadableType.Boolean -> array.put(readableArray.getBoolean(i))
+                ReadableType.Number -> array.put(readableArray.getDouble(i))
+                ReadableType.String -> array.put(readableArray.getString(i))
+                ReadableType.Map -> array.put(convertMapToJson(readableArray.getMap(i)))
+                ReadableType.Array -> array.put(convertArrayToJson(readableArray.getArray(i)))
+            }
+        }
+        return array
+    }
+
+    @Throws(JSONException::class)
+    private fun convertJsonToMap(jsonObject: JSONObject): WritableMap {
+        val map: WritableMap = WritableNativeMap()
+        val iterator = jsonObject.keys()
+        while (iterator.hasNext()) {
+            val key = iterator.next()
+            when (val value = jsonObject[key]) {
+                is JSONObject -> map.putMap(key, convertJsonToMap(value))
+                is JSONArray -> map.putArray(key, convertJsonToArray(value))
+                is Boolean -> map.putBoolean(key, value)
+                is Int -> map.putInt(key, value)
+                is Double -> map.putDouble(key, value)
+                is String -> map.putString(key, value)
+                else -> map.putString(key, value.toString())
+            }
+        }
+        return map
+    }
+
+    @Throws(JSONException::class)
+    private fun convertJsonToArray(jsonArray: JSONArray): WritableArray {
+        val array: WritableArray = WritableNativeArray()
+        for (i in 0 until jsonArray.length()) {
+            when (val value = jsonArray[i]) {
+                is JSONObject -> array.pushMap(convertJsonToMap(value))
+                is JSONArray -> array.pushArray(convertJsonToArray(value))
+                is Boolean -> array.pushBoolean(value)
+                is Int -> array.pushInt(value)
+                is Double -> array.pushDouble(value)
+                is String -> array.pushString(value)
+                else -> array.pushString(value.toString())
+            }
+        }
+        return array
+    }
+
+    private fun cleanup() {
+        localAssociation = null
+        mobileWalletAdapterClient = null
+        semTerminated = null
+        if (mutex.isLocked) {
+            mutex.unlock()
+        }
+    }
+
+    @ReactMethod
+    fun startSession(config: ReadableMap?, promise: Promise) = launch {
+        Log.d(name, "startSession with config $config")
+        mutex.lock()
+        try {
+            val semConnectedOrFailed = Semaphore(1, 1)
```

Comment:
We will be updating the API around this connection logic so a semaphore isn't needed.